source/slang/slang-ir-metal-legalize.cpp

```diff
@@ -1285,16 +1341,85 @@ namespace Slang
                     auto paramVarLayout = varLayoutBuilder.build();
                     builder.addLayoutDecoration(param, paramVarLayout);
                 }
+                if(param->findDecorationImpl(kIROp_VerticesDecoration))
+                {
+                    auto vertexRefType = as<IRPtrTypeBase>(param->getDataType());
+                    auto vertexOutputType = as<IRVerticesType>(vertexRefType->getValueType());
+                    vertexType = vertexOutputType->getElementType();
+                    maxVertices = vertexOutputType->getMaxElementCount();
+                    SLANG_ASSERT(vertexType);
+
+                    verticesParam = param;
+                    auto vertStruct = as<IRStructType>(vertexType);
+                    for(auto field : vertStruct->getFields())
+                    {
+                        auto key = field->getKey();
+                        if(auto deco = key->findDecoration<IRSemanticDecoration>())
+                        {
+                            if(deco->getSemanticName().caseInsensitiveEquals(toSlice("sv_position")))
+                            {
+                                builder.addTargetSystemValueDecoration(key, toSlice("position"));
+                            }
+                        }
+                    }
+                }
+                if(param->findDecorationImpl(kIROp_IndicesDecoration))
+                {
+                    auto indicesRefType = (IRConstRefType*)param->getDataType();
+                    auto indicesOutputType = (IRIndicesType*)indicesRefType->getValueType();
+                    indicesType = indicesOutputType->getElementType();
+                    maxPrimitives = indicesOutputType->getMaxElementCount();
+                    SLANG_ASSERT(indicesType);
+
+                    indicesParam = param;
+                }
+                if(param->findDecorationImpl(kIROp_PrimitivesDecoration))
+                {
+                    auto primitivesRefType = (IRConstRefType*)param->getDataType();
+                    auto primitivesOutputType = (IRPrimitivesType*)primitivesRefType->getValueType();
+                    primitiveType = primitivesOutputType->getElementType();
+                    SLANG_ASSERT(primitiveType);
+
+                    primitivesParam = param;
+                    auto primStruct = as<IRStructType>(primitiveType);
+                    for(auto field : primStruct->getFields())
+                    {
+                        auto key = field->getKey();
+                        if(auto deco = key->findDecoration<IRSemanticDecoration>())
+                        {
+                            if(deco->getSemanticName().caseInsensitiveEquals(toSlice("sv_primitiveid")))
```

Comment:
Is this logic here to workaround the bug?

Reply:
this is for annotating mesh and primitive attributes at all. Since they are not part of the return type, they dont get processed in `wrapReturnValueInStruct`. The other issue is when taking the a SV_PrimitiveID as an input parameter in the fragment shader.

Reply:
We probably also want to include a check to make sure PrimitiveType is a struct, but this can be done in a common verification step around slang-lower-to-ir.cpp::generateIRForTranslationUnit (line 11192) and doesn't need to be part of this PR.